CUDA triangular solver

msimberg opened the pull request:
This builds on top of #315 (for fixes made to HPX since 1.5.0), and #312 (for global CUDA/cuBLAS pools/executors), so the diff is quite massive.

~Still WIP, but~ here are the main changes:
- the blas tile functions now have callable object wrappers for easier use in generic contexts
- the wrappers (together with `unwrapExtendTiles` from #312) are used in all algorithms, also the CPU-only ones
- the triangular solver is now implemented generically for CPU and GPU (but could still be specialized for a backend if needed)
- umpire is used as a memory pool for cuda allocations (both pinned host and device memory)

msimberg commented on CMakeLists.txt:
```diff
@@ -41,6 +41,9 @@ if (DLAF_WITH_CUDA)
   find_package(CUDALIBS REQUIRED)
 endif()
 
+# TODO: use option?
```
Yes. See #317.

msimberg commented on CMakeLists.txt:
```diff
@@ -61,13 +64,19 @@ else()
 find_package(lapackpp REQUIRED)
 
+# ----- UMPIRE
+set(DLAF_WITH_UMPIRE OFF CACHE BOOL "Enable Umpire support")
```
Use `option`.

msimberg commented on include/dlaf/communication/sync/broadcast.h:
```diff
@@ -54,41 +55,98 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
       "recv_tile");
-  return hpx::dataflow(ex, std::move(recv_bcast_f), mpi_task_chain());
+
+  return detail::handle_recv_tile<D>::call(hpx::dataflow(ex, std::move(recv_bcast_f), mpi_task_chain()));
```
Relevant for #318.

msimberg commented on include/dlaf/tile.h:
```diff
@@ -0,0 +1,26 @@
+//
```
Rename to `include/dlaf/blas.h`?

albestro replied:
Do you see any plus in keeping it separate? Otherwise, I would go for keeping them all together.

msimberg replied:
No reason other than that the blas and cublas functionality have so far stayed mostly in separate files. This is the current `cublas_tile.h`: https://github.com/eth-cscs/DLA-Future/blob/e70c6518459bf1a0159d94660156e23380a07bf8/include/dlaf/cublas_tile.h. I could move the cublas declarations to `blas_tile.h` as well, and move the cublas definitions to `blas_tile.tpp`. In that case, should things like `util_blas.h` and `util_cublas.h` also be combined?

rasolca replied:
What about merging `include/dlaf/tile.h``include/dlaf/blas_tile.h` and `include/dlaf/cublas_tile.h` in a single file `include/dlaf/blas/tile.h`?
I would anyway keep `util_blas.h` and `util_cublas.h` separate and move them in `include/dlaf/blas` as well.

msimberg replied:
> What about merging `include/dlaf/tile.h`, `include/dlaf/blas_tile.h`, `include/dlaf/blas_tile.tpp` and `include/dlaf/cublas_tile.h` in a single file `include/dlaf/blas/tile.h`?

Ok with me. @albestro?

> I would anyway keep `util_blas.h` and `util_cublas.h` separate and move them in `include/dlaf/blas` as well.

Separately into `dlaf/blas/util.h` and `dlaf/cublas/util.h`?

I suppose there are related renamings that could be done for `lapack_tile.h`, `util_math.h`, `util_matrix.h`, and `util_tile.h`, but I won't move those in this PR.

msimberg commented on miniapp/miniapp_triangular_solver.cpp:
```diff
@@ -12,17 +12,23 @@
 
+#ifdef DLAF_WITH_CUDA
+#include <cublas_v2.h>
+#endif
```
This shouldn't be needed (anymore).

msimberg commented on spack/packages/dla-future/package.py:
```diff
@@ -25,9 +25,10 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("hpx cxxstd=14 networking=none")
-    depends_on("hpx@1.5.0", when="~cuda")
-    depends_on("hpx@master +cuda", when="+cuda")
+    depends_on("hpx+cuda", when="+cuda")
```
This should be changed back to 1.6.0 when it's out.

msimberg commented:
bors try

msimberg commented on CMakeLists.txt:
```diff
@@ -23,7 +23,12 @@ if (NOT CMAKE_BUILD_TYPE)
 option(BUILD_SHARED_LIBS "Build shared libraries." OFF)
+option(DLAF_WITH_CUDA "Enable CUDA support" OFF)
+cmake_dependent_option(DLAF_WITH_CUDA_MPI_RDMA "Enable MPI CUDA RDMA" OFF "DLAF_WITH_CUDA" OFF)
+cmake_dependent_option(DLAF_WITH_UMPIRE "Enable Umpire support" ON "DLAF_WITH_CUDA" OFF)
```
Umpire is mostly useless with the CPU backend. However, it also doesn't seem to hurt, and with spack, having it as a dependency is rather trivial. My preference would be to actually make it a required dependency since it avoids having to run tests with and without it. On the other hand, I see why it might be nice to not force it as a dependency (although for CUDA it's practically a must if one wants good performance).

Preferences?

teonnik replied:
My guess is that DLA-F will be built with CUDA support in production much more often than not. If that is indeed the case, I don't mind making Umpire a required dependency.

msimberg replied:
I'll make it required in that case.

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
Well, the test steps failed, but this is ready for a(nother) review now. @haampie/others is the gitlab instance known to not yet be fully functional after the upgrade?

All the tile blas operations are callable objects and should be used together with the helper `unwrapExtendTiles` from the previous PR. Umpire is still optional, but I'll change it to required if everybody thinks that's ok. The CUDA build in CI has it enabled and the CPU build has it disabled so it's decently covered, but the CUDA+no-umpire and CPU+umpire combinations are not built.

For the case when CUDA RDMA is not available, there are two helper functions `handleRecvTile` and `prepareSendTile` that copy the tiles in the GPU version. In the CPU version they are no-ops. In the terminology of https://github.com/eth-cscs/DLA-Future/pull/318#issue-555566711 these are "helpers" because of the need to specialize them into no-ops when the source and destination devices are the same. They are also now explicitly separate from the `send/recvTile` functions which will hopefully mean no/few conflicts with #228.

The `unwrapExtendTiles/getUnwrapReturnValue` combo is something that I really wish to see go away with senders/receivers (unwrapping becomes unnecessary, which in turn makes it easier for executors to generically just keep arguments alive longer if necessary), but they continue to be necessary at the moment.

Besides feedback on the functionality itself I'd appreciate comments on naming. E.g. is the helper `CommunicationDevice` confusingly named? I can see how it could be, but given that the place where memory is allocated is a `Device` it's not far from the truth.

rasolca commented:
bors try
(It was a scratch cleanup problem.)

albestro reviewed (changes requested):
In general, LGTM. Very nice work.

msimberg commented on include/dlaf/communication/sync/broadcast.h:
```diff
@@ -55,10 +57,29 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
 }
 }
 
+/// Helper function for preparing a tile for sending.
+///
+/// Duplicates the tile to CPU memory if CUDA RDMA is not enabled for MPI.
+/// Returns the tile unmodified otherwise.
+template <Device D, typename T>
+auto prepareSendTile(hpx::shared_future<matrix::Tile<T, D>> tile) {
+  return matrix::duplicateIfNeeded<CommunicationDevice<D>::value>(std::move(tile));
+}
+
+/// Helper function for handling a tile after sending.
+///
+/// If CUDA RDMA is disabled, the tile returned from recvTile will always be on
+/// the CPU. This helper duplicates to the GPU if the first template parameter
+/// is a GPU device. The first template parameter must be given.
+template <Device D, typename T>
+auto handleRecvTile(hpx::future<matrix::Tile<T, CommunicationDevice<D>::value>> tile) {
+  return matrix::duplicateIfNeeded<D>(std::move(tile));
+}
+
```
I agree! There's no particular reason for them to not be in `internal`. The main reason they're explicitly called inside the algorithm, rather than somewhere hidden is that the "wrapper" send/recv functions that used to call dataflow were lifted up so that dataflow is called explicitly in the algorithm. The only restriction is that `handleRecvTile/prepareSendTile` should ideally be outside the dataflow call.

`handleRecvTile` could in principle be inside `recvTile` as a `return dataflow(copyexecutor, handleRecvTile, ...)` at the end, but the future from `prepareSendTile` should really be passed to `dataflow(mpiexecutor, sendTile, ...)` instead of being called inside `sendTile`

I'm open to what you all think is clearest, but it sounds to me right now like I should reintroduce the wrapper functions that call dataflow.

albestro replied:
After your offline explanation of the restrictions I see your point.

IMHO in general we should try to prefer "schedulable" functions instead of "scheduling" ones (for reference see https://github.com/eth-cscs/DLA-Future/pull/318#issuecomment-767350611).

I think that with the new needs  of `handle/prepare` for communications, plus the fact that we always want to use the same executor for communication (right?), it is reasonable to make them "scheduling" functions, even if we will have to investigate if there is an easy way to customize annotation for them (e.g. "send diagonal tile row-wise" instead of a generic "send tile").

@msimberg @rasolca @teonnik from an API point of view, should we make it clear that they inside call `dataflow`, so they shall not go into a task?

**TL;DR** I agree you need to reintroduce the wrapper functions that calls dataflow (what I refer to as "scheduling" functions).

msimberg replied:
Well, I renamed the wrappers to `scheduleSendTile`/`scheduleRecv(Alloc)Tile` now. Alternatively I'd simply call them `sendTileWrapper` etc. I'm sure we'll come up with a better naming scheme in the future. Some potential alternatives:

- Move the "task" functions to the `internal` namespace to indicate that they shouldn't be called directly and call the wrappers plain `recv/sendTile`. I didn't move them now since they are still used in non-GPUized algorithms. The fact that the wrappers take an executor as the first argument is a good indication that they shouldn't be used directly as tasks.
- Somehow indicate the RDMA-ness of the communication functions. This is in principle orthogonal to whether or not they should be tasks in themselves, but may help clarify how they should be used. E.g. something like `recvTileRDMA` for the plain version that always receives on the target device and `recvTileViaCPU` for  a version that receives via the CPU (conditionally or not?).

Unless someone gets a big a-ha moment from these ideas I suggest we go ahead with something relatively generic like the `schedule*` or `*Wrapper` names.

msimberg commented on include/dlaf/communication/sync/broadcast.h:
```diff
@@ -55,10 +57,29 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
+template <typename T, Device D, template <class> class Future>
 void sendTile(hpx::future<common::PromiseGuard<comm::CommunicatorGrid>> mpi_task_chain, Coord rc_comm,
-              hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile) {
+              Future<matrix::Tile<T, D>> tile) {
```
It might be ok to make it `shared_future`, but I'd prefer to avoid unnecessarily converting everything to a `shared_future` if not necessary.

albestro replied:
Can I ask you the rationale? Because usually I always tried to ensure const-correctness, while it seems there is a more important reason for not doing so.

msimberg replied:
After discussing with @albestro I understood that I had read his question sloppily. I understood it to be about the change to have a generic `Future` type, rather than about the constness of `T`. It can and should indeed be `const` and I've just pushed a change which makes it `const`. Thanks @albestro for pointing this out!.

msimberg commented on include/dlaf/matrix/copy_tile.h:
```diff
@@ -98,25 +99,76 @@ struct CopyTile<T, Device::GPU, Device::GPU> {
 
 DLAF_MAKE_CALLABLE_OBJECT(copy);
+
+/// Helper struct for copying a given tile to an identical tile on Destination.
```
Yeah, identical means same type, same contents. Only the device might be different.

msimberg commented on include/dlaf/solver/triangular/impl.h:
```diff
@@ -390,11 +381,13 @@ void Triangular<Backend::MC, Device::CPU, T>::call_LLN(comm::CommunicatorGrid gr
 
         kk_tile = mat_a.read(kk);
-        dataflow(executor_mpi, comm::sendTile_o, serial_comm(), Coord::Row, mat_a.read(kk));
+        auto comm_tile = comm::prepareSendTile(mat_a.read(kk));
```
Agree, replied in more detail on the other comment.

msimberg commented on miniapp/miniapp_triangular_solver.cpp:
```diff
@@ -41,8 +43,15 @@ using dlaf::comm::Communicator;
 
-using T = double;
+using T = float;
+#if DLAF_WITH_CUDA
+constexpr Backend B = Backend::GPU;
+using HostMatrixType = dlaf::Matrix<T, Device::CPU>;
+using MatrixType = dlaf::Matrix<T, Device::GPU>;
+#else
+constexpr Backend B = Backend::MC;
 using MatrixType = dlaf::Matrix<T, Device::CPU>;
+#endif
```
I think it should eventually just be a command line option to the miniapps.

msimberg replied:
I suppose this could also be handled in a separate PR?

rasolca replied:
I agree to have an option for it and that it can be handled in a separate PR.

msimberg replied:
Opened https://github.com/eth-cscs/DLA-Future/issues/345 as a reminder to change it afterwards.

msimberg commented on spack/packages/dla-future/package.py:
```diff
@@ -25,6 +25,8 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("blaspp")
     depends_on("lapackpp")
+    depends_on("umpire~examples")
+    depends_on("umpire+cuda~shared", when="+cuda")
```
No. It should be disabled for cuda as well with that specification unless I'm missing something.

albestro replied:
my fault, you're right!

msimberg commented:
bors try

msimberg commented:
@albestro I've addressed some of your comments. I didn't yet do anything about
- the communication helpers (I moved them to `internal` but did not move the calls)
- did not move/merge any header files
I'd appreciate yours (and others') feedback on those before I do anything about them.

msimberg commented:
bors try

rasolca commented on include/dlaf/matrix/tile.h:
```diff
@@ -222,11 +223,9 @@ struct UnwrapFuture<hpx::future<Tile<T, D>>> {
+class UnwrapExtendTiles {
   template <typename... Ts>
-  auto operator()(Ts&&... ts) {
+  auto call_helper(std::true_type, Ts&&... ts) {
```
According to the style:
```suggestion
  auto callHelper(std::true_type, Ts&&... ts) {
```

rasolca reviewed (approved):
Looks good.
Just spotted a minor naming issue.

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
bors try

teonnik commented on include/dlaf/init.h:
```diff
@@ -48,6 +50,12 @@ struct Init {
   static void finalize() {}
 };
 
+template <>
+void Init<Backend::MC>::initialize(configuration const&);
+template <>
+void Init<Backend::MC>::finalize();
+extern template struct Init<Backend::MC>;
+
```
I forgot to add a comment, so I  will do it here: 

LGTM overall. 

The triangular solver has to be adapted to the new MPI API. The implementation of the MPI executor allows to be compatible with `getMPIExecutor<>()` API.

msimberg replied:
> Do you need this in the header for testing purposes?

Good question, I don't remember (it wasn't for testing purposes as far as I can remember). It'd indeed be nice if it can go in the cpp file. Will check again.

> I forgot to add a comment, so I will do it here:
> 
> LGTM overall.
> 
> The triangular solver has to be adapted to the new MPI API. The implementation of the MPI executor allows to be compatible with `getMPIExecutor<>()` API.

Yeah, I'm changing that right now. Given that the executor's constructor now chooses the correct pool `getMPIExecutor` is somewhat unnecessary. It might be worth keeping it just for symmetry with the other executors though.

Did I understand it correctly that one doesn't actually need a separate row and column executor (they're exactly the same)? I presume this was from trying to have the executor know which type it is (which we should still look at)?

teonnik replied:
> Did I understand it correctly that one doesn't actually need a separate row and column executor (they're exactly the same)? I presume this was from trying to have the executor know which type it is (which we should still look at)?

Yes, the row/column distinction comes from the communicator which is now handled outside of the executor. The only exception is if we want to later bring back "blocking" MPI version of the executor. In that case we would need to save the index of the OS thread to which the executor is assigned. That means we would need two different executors for the row and column communicators. However, as we decided to remove the "blocking" version, a single executor is sufficient.

msimberg replied:
> Do you need this in the header for testing purposes?

It was unnecessary, I've moved it to the cpp file.

For the triangular solver I now just use one executor for both column and row communication. It should be easy enough to change it back when/if needed.

albestro replied:
I agree with you about having a single Executor for both row and col...

I have one more question: why not having a single Executor also for all MPI communications? I mean, for the moment is not a big deal because it does not take any parameter, but it may be reasonable to have it inside our `initialize` and get it via the `getMPIExecutor()`. Do you see any problem? This may be the topic of a dedicated issue/PR anyway...

msimberg replied:
My wording was maybe confusing. The same executor is used for both types of communication now. If you're talking about using literally the same instance that's also a possibility (i.e. returning a reference to a function static executor); currently it's a new instance every time you call `getMPIExecutor`, but I don't see that as a huge overhead (executors are cheap to create).

msimberg commented on spack/packages/dla-future/package.py:
```diff
@@ -25,7 +25,9 @@ class DlaFuture(CMakePackage, CudaPackage):
-    depends_on("hpx cxxstd=14 networking=none")
+    depends_on("umpire~examples")
+    depends_on("umpire+cuda~shared", when="+cuda")
+    depends_on("hpx cxxstd=14 networking=none +async_mpi")
```
NB. I added `+async_mpi` here. If someone wants it I can make a separate PR for that.

msimberg commented:
bors try

msimberg commented:
This seems to pass tests locally now after the merge with master.

I made one additional change to the header file organization: I moved the triangular solver declarations required by `triangular.h` to `triangular/api.h`, and `triangular/impl.h` is now never included in user-code, since it's not necessary (only necessary for the instantiations). This means that changes to the triangular solver implementation doesn't cause tests/examples using it to be recompiled. Let me know if you think this sounds reasonable, or if I should revert and we handle this separately.

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented:
bors try

rasolca commented:
bors try
(codecov checksum should be fixed)

msimberg commented:
bors try

msimberg commented:
The one failure looks to me unrelated to the changes here: https://gitlab.com/cscs-ci/eth-cscs/DLA-Future/-/jobs/1202885692. It seems to have successfully run the tests but was stuck on creating the coverage reports.

rasolca commented:
It is not stuck, but it takes long time to initialize the report and to create it and we often hit the time limit.
I'll check if I find a way to speed things up, otherwise I'll increase the time limit.

msimberg commented:
bors try

albestro reviewed (approved):
LGTM
